airflow/providers/google/cloud/hooks/bigquery_reservation.py

```diff
@@ -0,0 +1,125 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""This module contains a BigQuery Reservation Hook."""
+from __future__ import annotations
+
+from typing import Sequence
+
+from google.cloud.bigquery_reservation_v1 import ReservationServiceClient
+
+from airflow.exceptions import AirflowException
+from airflow.providers.google.common.consts import CLIENT_INFO
+from airflow.providers.google.common.hooks.base_google import GoogleBaseHook
+
+
+class BigQueryReservationServiceHook(GoogleBaseHook):
+    """
+    Hook for Google Bigquery Reservation API.
+
+    :param gcp_conn_id: The Airflow connection used for GCP credentials.
+    :param location: The location of the BigQuery resource.
+    :param impersonation_chain: This is the optional service account to impersonate using short term
+        credentials.
+    """
+
+    conn_name_attr = "gcp_conn_id"
+    default_conn_name = "google_cloud_bigquery_reservation_default"
+    hook_name = "Google Bigquery Reservation"
+
+    def __init__(
+        self,
+        gcp_conn_id: str = GoogleBaseHook.default_conn_name,
+        location: str | None = None,
+        impersonation_chain: str | Sequence[str] | None = None,
+    ) -> None:
+        super().__init__(
+            gcp_conn_id=gcp_conn_id,
+            impersonation_chain=impersonation_chain,
+        )
+        self.location = location
+        self._client: ReservationServiceClient | None = None
+
+    @staticmethod
+    def _convert_gb_to_kb(value: int) -> int:
+        """
+        Convert GB value to KB.
+
+        :param value: Value to convert
+        """
+        return value * 1073741824
+
+    def get_client(self) -> ReservationServiceClient:
+        """
+        Get reservation service client.
+
+        :return: Google Bigquery Reservation client
+        """
+        if not self._client:
+            self._client = ReservationServiceClient(
+                credentials=self.get_credentials(), client_info=CLIENT_INFO
+            )
+        return self._client
+
+    @GoogleBaseHook.fallback_to_default_project_id
+    def create_bi_reservation(self, project_id: str, size: int) -> None:
```

Comment:
do we need project_id args here? can we access this with `self.project_id` since this is a base class property?

Reply:
Most of the time you wish to create bi_engine reservation not on the same project than the default project (defining in the connection). 
I choose to put it as the function argument but we could add it in the `BigQueryReservationServiceHook` attribute. What do you prefer ?

Reply:
hmm, in that case I would prefer to have this in `BigQueryReservationServiceHook`

Reply:
I try to do like this but currently it is not possible to define `project_id` as an attribute. 
This is because of `GoogleBaseHook` have defined `project_id` with a decorated @property and it does not have a setter method for `project_id` ([see](https://github.com/apache/airflow/blob/main/airflow/providers/google/common/hooks/base_google.py#L342)).  

I am not confortable to change it in this PR.

Reply:
I take a look to other Google hook, most of them have`project_id` as an arguments of functions. 
Currently, I propose to you to stay like this. 
But, perahps open an issue and a new PR to change this. 
Do you agree ?

Reply:
I see - then I'm ok to keep it in its current form. thanks!